charmcraft/config.py

```diff
@@ -105,6 +114,73 @@ def from_dict(cls, source):
         return cls(**source)
 
 
+@attr.s(kw_only=True, frozen=True)
+class BaseConfig:
+    """Configuration for all Charmhub related options."""
+
+    name = attr.ib(validator=attr.validators.instance_of(str))
+    channel = attr.ib(validator=attr.validators.instance_of(str))
+    architectures = attr.ib(default=attr.Factory(list), validator=attr.validators.deep_iterable(
+        member_validator=attr.validators.instance_of(str),
+        iterable_validator=attr.validators.instance_of(list)))
+
+    @classmethod
+    def from_dict(cls, source):
+        """Build from a raw dict."""
+        # Ensure channel is treated as string if interpreted as number, e.g. 20.04.
+        channel = str(source.pop("channel"))
+
+        return cls(**source, channel=channel)
+
+
+@attr.s(kw_only=True, frozen=True)
+class BuildonRunonBaseConfig:
+    """Configuration for all Charmhub related options."""
+
+    build_on = attr.ib(default=attr.Factory(list))
+    run_on = attr.ib(default=attr.Factory(list))
+
+    @classmethod
+    def from_dict(cls, source):
+        """Build from a raw dict."""
+        build_on = list()
+        run_on = list()
+
+        if "build-on" in source or "run-on" in source:
+            # Long-form syntax for explicit build-on/run-on scenarios.
+            for build_base in source.get("build-on", list()):
+                build_on.append(BaseConfig.from_dict(build_base))
+
+            for run_base in source.get("run-on", list()):
+                run_on.append(BaseConfig.from_dict(run_base))
```

Comment:
Two details I think are not handled in this logic:

- if the config says "run-on" but not "build-on" it should be an error.
- if the config says "build-on" but not "run-on", the later should default to the former (not being empty).

Another detail: it's faster and simpler to initialize an empty list using `[]` than calling the function.